---
title: "Lung_carcinoid_analysis_github_upload"
author: "Christin"
date: "11/28/2022"
output: html_document
---
#Data tables were generated by using published data from Alcala et al, 2019; Cuesta et al, 2014; Laddha et al, #2019 and data files were downloaded from #https://nextjournal.com/rarecancersgenomics/a-molecular-map-of-lung-neuroendocrine-neoplasms/
#Lungs desingated as typical or atypical carcinoids based on histopathology were selected for this analysis. Data #were normalized using quantile normalization (Dillies et al, 2013) and the relevant neuropeptide, hormone, and marker genes are represented in the data tables uploaded to this github site.

#Import data files for all tumors with normalized values
```{r}
LNET_all <- read.table('LNETs_Normalized_all.csv', header = T, sep = "," )
#assign row.names
row.names(LNET_all) <- LNET_all[,1]
```
#Subset carcinoids
```{r}
#define vector of sample IDs corresponding to typical carcinoids defined by histopathology

Typical <- c("S00016","S00076","S00089", "S00094", "S00118","S00128", "S00515","S00516","S00520", "S00858","S01060","S01103","S01493","S01501","S01502","S01504", "S01510", "S01515", "S01519", "S01521","S01526","S01528", "S01529", "S01531", "S01532", "S01536", "S01537", "S01538", "S01539", "S01540", "S01541", "S01543", "S01545","S01546", "S01567","S01573","S01583","S01584", "S01585", "S01590","S01593", "S01605", "S01666", "S01731", "S01733", "S01742", "S01746","S02325","S02326","S02327","S02330","S02331", "S02333","S02334", "S02335", "S02337","S02338","S02339","SRR7646229","SRR7646235","SRR7646236","SRR7646237","SRR7646239","SRR7646242","SRR7646247","SRR7646248","SRR7646249","SRR7646250","SRR7646251","SRR7646252", "SRR7646232","SRR7646253","SRR7646254","SRR7646256")

#define vector of sample IDs corresponding to atypical carcinoids defined by histopathology
Atypical <- c("LNEN002","LNEN003","LNEN004","LNEN005","LNEN006","LNEN007","LNEN008","LNEN009","LNEN010","LNEN011","LNEN012","LNEN013","LNEN014","LNEN015","LNEN016","LNEN017","LNEN018","LNEN019","LNEN020", "S01202","S01513","S01520","S01572","S02340","SRR7646230","SRR7646231","SRR7646233","SRR7646234","SRR7646238","SRR7646240","SRR7646243","SRR7646244","SRR7646245","SRR7646246","SRR7646255","SRR7646257","SRR7646258")
```
#subset data for typical and atypical carcinoids
```{r}
typical_data <- LNET_all[,Typical]
atypical_data <- LNET_all[,Atypical]
```
#Figure S13 plots
```{r}
#define genes to feature (markers, neuropeptides)
markers_ordered <- c('SCGN','DIO2','CPLX2','DIRAS3','ASCL1','GP2','SLC35D3','FXYD2','NCAM1','SYP','INSM1')
#ordered human NP genes detected in scRNA-seq human PNECs
Hu_NP_genes <- c('SCG5','PCSK1N','GRP','CHGA','SCG2','SCG3','CALCA','CHGB','DBI','POMC','TRH','ADCYAP1','VGF','INHBA','CGA','AMH','INHBB','NPPA','NPW','CALCB','NMB','ADM','SST','INHA','CARTPT','GHRL','AGT','HCRT','CCK','AGRP','PRLH','NPY','EDN1','GNRH2','KNG1','UCN','UTS2B','GNRH1','EPO','REN')

#mouse-specific NPs. 
mouse_NPs <- c('IAPP','CRH','EDN3','NPFF','NPB','UTS2','UCN2','PNOC','OXT','PROK2','IGF1','IGF2','LHB','NPPB','NPPC')

non_detect <- c('PENK','TAC1','VIP', 'ADIPOQ','APLN','AVP','CORT', 'EDN2','FSHB','GALP','GAST', 'GCG','GIP','HAMP','INS','LEP','MLN','NMS','NMU','NPVF','NTS','OSTN','PDYN','PMCH','PRL','PTH','PTHLH','PYY', 'RETN','RLN1', 'SCT')

#Gene list for final plots
all_genes <- c(markers_ordered, Hu_NP_genes, 'GHRH',mouse_NPs, non_detect)
```

#subset genes

```{r}
typical_data_genes <- LNET_all[c(markers, Hu_NP_genes, mouse_NPs, non_detect), Typical]
```
#convert to data matrix
```{r}
typical_data <- data.matrix(typical_data)
```
#transpose
```{r}
data_matrix <- t(typical_data)
```
#plot
```{r}
library(pheatmap)
pheatmap(data_matrix, cluster_rows = FALSE, legend = TRUE, cluster_cols = FALSE,
show_colnames = TRUE, show_rownames = TRUE,  fontsize_row = 6)
```



