---
title: "Mouse notebook for scRNAseq analysis of PNECs"
author: "Christin Kuo"
date: "4/12/2021"
output: html_document
---

Load libraries
```{r setup, include=FALSE}
library(useful)
library(Seurat)
library(dplyr)
library(Matrix)
library(ontologyIndex)
require(corrplot)
require(GGally)
require(reshape)
require(gplots)
library(here)
library(plyr)
library(RColorBrewer)
library(tidyr)

source('HelperFunctions.R')
my.cols <- rev(brewer.pal(11, "RdYlBu"))
```
#specify directory
```{r}
dir <- "/Users/ckuo/Documents/Active projects/Single Cell Manuscript/Mouse_NE_2020/Mouse_PNEC_20210407"
```


Removing doublets from the dataset using Scrublet
Load master gene counts table
```{r}
data_adult <- read.table('GC_table_PN_NE_all_manuscript.tsv')
#metadata for all cells
meta_adult_new <- read.table('Metadata_table_postnatal_all.tsv')
```
#Get all experiment IDs and plot a table
```{r Separate experiment IDs}
exps <- unique(meta_adult_new$experiment.ID)
table(meta_adult_new$experiment.ID)
```
#Generating matrix 
```{r Divide data by experiment and write matrix to Scrublet output}
require(Matrix)
# Get all experiment IDs
# exps <- c(6,7,8)
list.names <- list()
for(i in 1:length(exps)){
  # Subset cells and adata for each experiment ID
  cells <- row.names(meta_adult_new)[which(meta_adult_new$experiment.ID==exps[i])]
  list.names[[i]] <- cells
  paste("Ncells for experiment ID", exps[i], "=", length(cells))
  message(paste("Ncells for experiment ID", exps[i], "=", length(cells)))
  data.temp <- as.matrix(data_adult[,cells])
  # Convert to sparse format 
  data.temp <- Matrix(data.temp, sparse=T)
  colnames(data.temp) <- cells
  
#write table to input folder
  writeMM(data.temp, file=paste(paste(dir,"/input/",sep=""),"Exp.ID.", exps[i],".mtx" ,sep=""))
  }
rm(data.temp)
```

#Code to run 'NE_scrublet_basics' to obtain doublet score for each cell in each run. See github files, and once Scrublet run, use this version of data table for subsequent analysis. 

Load data for tables with doublets removed already
```{r Load data}

data_adult <- read.table('GC_table_adult_NE_renamed_ND.tsv')

#loading in new metadata file. Prepared for manuscript.
meta_data <- read.csv('MetaData_mouse.csv', header = TRUE, row.names = 1)
```

Create Seurat object 
```{r create Seurat}
require(Seurat)
tiss <- CreateSeuratObject(raw.data = data_adult, meta.data = meta_data, 
                           display.progress = T)
tiss <- AddMetaData(object = tiss, metadata = meta_data)
# save(tiss, file = paste(dir,"/output/Mouse_Lung_Seurat.RData",sep=""))
```


Filter the dataset and preprocess 
Filtering on 1000 genes minimum and 50000 reads minimum per cell 
```{r Filter}
tiss <- FilterCells(object = tiss, subset.names = c("nGene", "nUMI"), 
    low.thresholds = c(1000, 50000), high.thresholds = c(Inf, Inf))
tiss
```

Preprocess data 
```{r echo = T, results = 'hide'}
tiss <- NormalizeData(object = tiss, normalization.method = "LogNormalize", 
    scale.factor = 1000000)
tiss <- ScaleData(object = tiss, do.scale = T, do.center = T)
```

Find variable genes
```{r Feature selection}
tiss <- FindVariableGenes(object = tiss, do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 1, num.bin = 20, binning.method = "equal_width", do.recalc = T, x.low.cutoff=1)
```

Perform PCA
```{r PCA analysis}
tiss <- RunPCA(object = tiss, do.print = FALSE)
tiss <- ProjectPCA(object = tiss, do.print = FALSE)
```


Visualize top genes across principal components
```{r Visualize top PCA genes}
PCHeatmap(object = tiss, pc.use = 1:15, cells.use = 100, do.balanced = TRUE, label.columns = FALSE, num.genes = 10)
```

Also visualize PC variance
```{r Variance per PC}
#comp.dev <-pca.obj@sdev
PCElbowPlot(object = tiss)
```

Choose the number of principal components to use.
```{r choose PCs}
n.pcs = 15
```

Set resolution and perform clustering
```{r choose resolution}
res.used <- 0.3
tiss <- FindClusters(object = tiss, reduction.type = "pca", dims.use = 1:n.pcs, 
    resolution = res.used, print.output = 0, save.SNN = TRUE, k.param = 5)
```

Perform  tSNE
```{r}
set.seed(123)
tiss <- RunTSNE(object = tiss, dims.use = 1:n.pcs, seed.use = 10, perplexity=30)
```

Visualize TSNE colored by cluster
```{r}
TSNEPlot(object = tiss, do.label = T)
```

Plot the fraction of each dataset in each
```{r}
tab.1 <- table(tiss@meta.data$plate,tiss@meta.data$res.0.3)
```


Color by a gene. Feature plot to visualize and identify PNECs
```{r}
FeaturePlot(tiss, c("Ascl1", "zsGreen_transgene","Calca", "Epcam"), pt.size = 1, nCol = 2, cols.use = my.cols)
```

Visualize TSNE metadata field
Experiment ID
```{r}
# note that you can set do.label=T to help label/visualize patterns by metadata field
TSNEPlot(object = tiss, do.label = F, group.by = "Age")
```

Save object 
```{r}
save(tiss,file=paste(dir, "/output/All_datasets_renamed_ND.RData",sep=""))
```
#Marker analysis
```{r}
tiss.markers <- FindAllMarkers(tiss, min.pct = 0.25, thresh.use = 0.25, only.pos = T)
#running markers against only epithelial population
tiss.epi.markers <- FindAllMarkers(subtiss_epi,min.pct = 0.25, thresh.use = 0.25, only.pos = T)
```
#To subset NE cells
```{r}
cells.to.use = WhichCells(tiss, c(0))
subtiss_NE = SubsetData(tiss, cells.use = cells.to.use) 
```
#Barplot of PNEC neuropeptide expression
```{r}
#Import data for NP analysis.
All_NP <- read.csv('All_NP_2020_10_18.csv')
NP_genes <- All_NP[,'MGNC']
NP_genes <- as.character(unique(NP_genes))
```
#To make fraction table for barplot using NP gene list

```{r}
subtiss_NE_NP <- FetchData(subtiss_NE, c(NP_genes))
subtiss_NE_NP <- t(subtiss_NE_NP)
NE_NP_frac <- round(rowSums(subtiss_NE_NP > 0) / dim(subtiss_NE_NP)[2],2)
```
#To make barplot ordered in the order: largest fraction of PNECs peptide in PNEC dataset #1, followed by similar ranking in PNEC#2.
```{r}
par(las=2)
barplot(NE_NP_frac, cex.names = 0.6, space = 0.1, ylim = c(0,1))
```
#Fig. 2 NP combinations
```{r}
#To make table of expressed NPs
subtiss_NE_NP_exp <- FetchData(subtiss_NE, c("Calca","Scg5", "Chgb","Scg2","Pcsk1n", "Chga","Scg3", "Cartpt", "Dbi", "Vgf","Agt","Calcb","Iapp","Adcyap1","Pomc","Adm","Agrp","Crh","Edn3","Gal","Gnrh1","Hcrt","Nmb","Npff","Nppa","Npw", "Uts2"))


```
#NEDotPlot function
```{r}
NEDotPlot <- function(barcodes, cols.use = c("lightgrey", "blue"),

          col.min = -2.5, col.max = 2.5, dot.min = 0, dot.scale = 6, scale.min = NA, scale.max = NA,

          group.by, plot.legend = FALSE, do.return = FALSE, x.lab.rot = FALSE, levels.order = FALSE, only.exp = FALSE)

{

  barcodes <- as.matrix(barcodes)

  barcodes <- barcodes > 0

  barcodes[which(barcodes == TRUE)] <- 1

  barcodes <- as.data.frame(t(barcodes))

  data.to.plot <- barcodes

  ids <- c()

  for (i in 1:dim(barcodes)[1]) {

    ids <- c(ids,paste(as.numeric(barcodes[i,]), collapse=""))

  }

  names(ids) <- rownames(barcodes)

  ids <- as.factor(ids)

  genes.plot <- colnames(barcodes)

  data.to.plot$cell <- rownames(x = data.to.plot)

  data.to.plot$id <- ids

  data.to.plot <- data.to.plot %>% gather(key = genes.plot, value = expression, -c(cell, id))

  if (only.exp == TRUE) {

    data.to.plot <- data.to.plot %>% group_by(id, genes.plot) %>%

      summarize(avg.exp = mean(log2(expm1(x = expression[expression > 0]) + 1)), pct.exp = PercentAbove(x = expression, threshold = 0))

  } else {

    data.to.plot <- data.to.plot %>% group_by(id, genes.plot) %>%

      summarize(avg.exp = mean(log2(expm1(x = expression) + 1)), pct.exp = PercentAbove(x = expression, threshold = 0))

  }

  data.to.plot <- data.to.plot %>% ungroup() %>% group_by(genes.plot) %>%

    mutate(avg.exp = MinMax(data = avg.exp, max = col.max, min = col.min))

  data.to.plot$genes.plot <- factor(x = data.to.plot$genes.plot,

                                    levels = sub(pattern = "-", replacement = ".",

                                                 x = genes.plot), ordered = TRUE)

 

  data.to.plot$pct.exp[data.to.plot$pct.exp <= dot.min] <- NA

  p <- ggplot(data = data.to.plot, mapping = aes(x = genes.plot,

                                                 y = id)) + geom_point(mapping = aes(size = pct.exp,

                                                                                     fill = avg.exp), colour="white",pch=21) + scale_radius(range = c(1,

                                                                                                                                                      

                                                                                                                                                      

                                                                                                                                                      dot.scale), limits = c(0,1)) + scale_fill_gradientn(colors = cols.use, guide = guide_colorbar(ticks = FALSE)) + theme(axis.title.x = element_blank(), axis.title.y = element_blank()) + coord_flip()

  if (!plot.legend) {

    p <- p + theme(legend.position = "none")

  }

  if (x.lab.rot) {

    p <- p + theme(axis.text.x = element_text(angle = 90,

                                              hjust = 1))

  }

  suppressWarnings(print(p))

  if (do.return) {

    return(p)

  }

}
```

#Subset and create objects of subsequent analysis
```{r}
#to create epithelial object
```{r}
cells.to.use = WhichCells(tiss, c(0,2,6,7,5,9))
subtiss_epi = SubsetData(tiss, cells.use = cells.to.use) 
```
#to create epithelial object from entire mouse lung cell atlas
```{r}
cells.to.use = WhichCells(tiss.mouse.anno, c(14,15,27,20,8,22))
subtiss_epi = SubsetData(tiss.mouse.anno, cells.use = cells.to.use) 
```
#20210515_Fig. 1B. to make marker analysis, compare to all epithelial populations in MCLA: cluster 14, NE; 15, ciliated; 27 basal; 20, AT1 27; AT2, 20; 8, ; 22
```{r}
tiss.markers.epi <- FindAllMarkers(subtiss_epi,min.pct = 0.25, thresh.use = 0.25, only.pos = T)

#selected top 100 genes for Table S1. For Plot in Panel B, used the first 20 genes> logfold change calculated from taking ratio NE vs. nonNE epithelial cells across all MCLA.
Log_fold_change <- c(7.68, 7.48, 7.19, 7.03, 6.8, 6.72, 6.53, 6.37, 6.31, 6.25, 6.17, 6.12, 6.08, 6.02, 5.98, 5.93, 5.76, 5.68, 5.64, 5.6)
barplot(Log_fold_change, cex.names = 0.6, space = 0.1, ylim = c(0,100), beside=TRUE, main = "Fig 1B_plot")

#5-17-21- values for genes log-fold change list
#Resp18 7.7, Pcsk1 7.5, Nov 7.2, Calca 7.0, Fam38b 6.8, Scg5 6.5, Col8a1 6.4, Spock3 6.3, Snap25 6.1, Tcerg1l 6.1, Ascl1 6.0, Ptprn 5.9, Arc 5.6, Nnat 5.4, Rab3b 4.7, Meg3 4.6, Cplx2 4.4, Slc35d3 4.3, Igf2 4.3, Pkib 2.8
Log_fold_change <- c(7.7,7.5,7.2,7,6.8,6.5,6.4,6.3,6.1,6.1,6,5.9,5.6,5.4,4.7,4.6,4.4,4.3,4.3,2.8)

barplot(Log_fold_change, cex.names = 0.6, space = 0.1, ylim = c(0,8), beside=TRUE, main = "Fig 1B")
```

#heatmap 
```{r}
library(pheatmap)

#cells_to_use <- WhichCells(tiss, c(0,2,5,6,7,9,1,3,4,8))
#cell order used in manuscript FigS6
cells_to_use <- WhichCells(tiss, c(0,2,6,7,5,9,3,1,4,8))
#epithelial cell clusters: 0,2,6,7,5,9
genes_to_use <- c(sensory_genes)
#genes_to_use = c('Car1','Car2','Car3','Car4','Car5a','Car5b','Car6','Car7','Car9','Car10','Car11','Car12','Car13','Car14','Ascl1','Pcsk1','Resp18','Calca')
annotation_column = data.frame("ident"= as.character(tiss@ident[cells_to_use]))
row.names(annotation_column) = cells_to_use
#if you want to do all cells: row.names(annotation_column) = rownames(tiss@data)

#data (log2(CPM+1)) --> better to use this one for heatmaps
pheatmap(tiss@data[genes_to_use,cells_to_use], cluster_rows = F, cluster_cols = F, show_colnames = F, annotation_col = annotation_column)

# you can choose to cluster the rows (genes) and/or the columns (cells)
```
#FigS6.Sensory heatmaps 20210502
#Mechanosensory
```{r}
mechano_genes <- read.csv("Mechano_genes.csv")

mechano_genes_list <- mechano_genes[,'MGNC']
mechano_genes_list <- as.character(unique(mechano_genes_list))

```
#Thermal_genes
```{r}
thermal_genes <- c('Trpa1','Trpv1','Trpv2','Trpv3','Trpv4','Trpm8','Grik2')
```
#Oxygen genes. 
```{r}
O2_genes <- c('Hif1a','Epas1','Hif3a','Egln1','Egln2','Egln3','Hif1an','Nox1','Nox3','Nox4','Cybb','Cyba','Ncf1','Ncf2','Rac2','Kcnk3','Kcnc3','Kcnc4','Kcnd3')
#mitochrondial hypoxia sensory genes
#mito_genes <- read.csv('genes_O2.csv', header = TRUE, sep = " ")
```

```{r}
mito_genes <- c('Ndufs1','Ndufs2','Ndufs3','Ndufs4','Ndufs5','Ndufs6','Ndufs7','Ndufs8','Ndufv1','Ndufv2','Ndufa1','Ndufa2', 'Ndufa3','Ndufa5','Ndufa6','Ndufa7','Ndufa8','Ndufa9','Ndufa10','Ndufa11','Sdha','Sdhb','Sdhc','Sdhd','Surf1','Uqcrb','Uqcrq','Uqcrc1','Uqcrc2','Uqcrh','Uqcr10','Uqcr11','Cyc1','Uqcrfs1','Cox5a','Cox5b')
genes_to_use <- mito_genes
```
#mito_genes_2 <- c('Ndufs4','Surf1')
```{r}
annotation_column = data.frame("ident"= as.character(tiss@ident[cells_to_use]))
row.names(annotation_column) = cells_to_use
```
```{r}
pheatmap(tiss@data[genes_to_use,cells_to_use], cluster_rows = F, cluster_cols = F, show_colnames = F, annotation_col = annotation_column)
```


#Acid genes
```{r}
acid_genes <- c('Accn2','Accn1','Accn3','Accn4','Kcnk3','Kcnk5','Kcnk9','Gpr4','Gpr132','Gpr68','Gpr65')
```
#CO2
```{r}
CO2_genes <- c('Car1','Car2','Car3','Car4','Car5a','Car5b','Car6','Car7','Car9','Car10','Car11','Car12','Car13','Car14')
```
#chemosensory
```{r}
chemo_genes <- c('Olfr107','Olfr1239','Olfr1243','Olfr1249','Olfr1413','Olfr292','Olfr453','Olfr494','Olfr555','Olfr592','Olfr613','Olfr615','Olfr71','Olfr738','Olfr843','Olfr90','Olfr919','Olfr92','Olfr93','Olfr78','Olfr558')
```
#phermone. Only expressed genes below (in 176 PNECs, mouse)
```{r}
phermone <- c('Vmn2r29','Vmn2r69')
```

#combined list for master heatmap
```{r}
sensory_genes <- c(mechano_genes_list, 'Mcoln3',thermal_genes, O2_genes, mito_genes, acid_genes, CO2_genes, chemo_genes, phermone)

genes_to_use <- sensory_genes
```

#Fig1G
```{r}
Fig1G_data <- read.csv('Fig1G_20210505_data.csv', header = TRUE, sep = ",")
Fig1G_data <- data.matrix(Fig1G_data)
Fig1G_data <- Fig1G_data[,2:9]
```
#plot barplot
```{r}
ISH_all <- Fig1G_data[1,]
barplot(ISH_all, cex.names = 0.6, space = 0.1, ylim = c(0,100), beside=TRUE, main = "ISH-all")

#adjusted for greater than or equal to 5 puncta
ISH_adj <- Fig1G_data[2,]
barplot(ISH_adj, cex.names = 0.6, space = 0.1, ylim = c(0,100), beside=TRUE, main = "ISH_adj")

#scRNA-seq data
RNA_seq <- Fig1G_data[3,]
barplot(RNA_seq, cex.names = 0.6, space = 0.1, ylim = c(0,100), beside=TRUE, main = "RNA-seq")
```

#Fig1. Ratio expression values NE vs. nonNE
```{r}
#make object for NE and non-NE
tiss_NE <- SubsetData(tiss, ident.use = 0)
tiss_nonNE <- SubsetData(tiss, ident.use = 1,2,3,4,5,6,7,8,9)
```
#Calculate 
```{r}
#To build proper data tabe (showing gene counts), need to subset the NE cells (using names stored in vector 'cells') and subset from original data file read into the seession called data.adult
#first define a vector name of cell names
cells <- rownames(tiss_NE@meta.data)
data_NE <- data_adult[, cells]
#now have a GCT of normal mouse NE
data_nonNE <- data_adult[,cells_nonNE]
```
#defining nonNE
```{r}
cells_nonNE <- rownames(tiss_nonNE@meta.data)
```
#calculating for NE
```{r}
nreads <- tiss_NE@meta.data[,'nUMI']
```

```{r}
gene.cpm <- (data_NE['Resp18',]/nreads * 10^6)
gene.cpm.1 <- gene.cpm + 1
#To get mean expression
#cell_num <- sum(gene.cpm > 0)
cell_num <- length(gene.cpm.1)
#to get mean expression
Avg_1 <- (sum(gene.cpm.1))/cell_num
#To take log
#Ln_cpm <- log2(Avg_1)
#To add to table. First define the first 2 rows of CPM_table <- rbind(gene.cpm.1, Ln_cpm) and add to that as follows with each gene change
```


#calculating for nonNE
```{r}
nreads_2 <- tiss_nonNE@meta.data[,'nUMI']
```

``
#Take ratio, then ln
```{r}
ratio <- log(Avg_1/Avg_2)
```
#add to table
```{r}
table_values <- rbind(table_values, ratio)
```


